searches/binary_search.py

```diff
@@ -1,161 +1,14 @@
-    if index != len(sorted_collection) and sorted_collection[index] == item:
-        return index
-    return None
-
-def binary_search_by_recursion(sorted_collection, item, left, right):
-
-    """Pure implementation of binary search algorithm in Python by recursion
-
-    Be careful collection must be sorted, otherwise result will be
-    unpredictable
-    First recursion should be started with left=0 and right=(len(sorted_collection)-1)
-
-    :param sorted_collection: some sorted collection with comparable items
-    :param item: item value to search
-    :return: index of found item or None if item is not found
-
-    Examples:
-    >>> binary_search_std_lib([0, 5, 7, 10, 15], 0)
-    0
-
-    >>> binary_search_std_lib([0, 5, 7, 10, 15], 15)
-    4
-
-    >>> binary_search_std_lib([0, 5, 7, 10, 15], 5)
-    1
-
-    >>> binary_search_std_lib([0, 5, 7, 10, 15], 6)
-
-    """
-    if (right < left):
-        return None
-    
-    midpoint = left + (right - left) // 2
-
-    if sorted_collection[midpoint] == item:
-        return midpoint
-    elif sorted_collection[midpoint] > item:
-        return binary_search_by_recursion(sorted_collection, item, left, midpoint-1)
-    else:
-        return binary_search_by_recursion(sorted_collection, item, midpoint+1, right)
-      
-def __assert_sorted(collection):
-    """Check if collection is sorted, if not - raises :py:class:`ValueError`
-
-    :param collection: collection
-    :return: True if collection is sorted
-    :raise: :py:class:`ValueError` if collection is not sorted
-
-    Examples:
-    >>> __assert_sorted([0, 1, 2, 4])
-    True
-
-    >>> __assert_sorted([10, -1, 5])
-    Traceback (most recent call last):
-    ...
-    ValueError: Collection must be sorted
-    """
-    if collection != sorted(collection):
-        raise ValueError('Collection must be sorted')
-    return True
-
-
-if __name__ == '__main__':
-    import sys
-    user_input = raw_input('Enter numbers separated by comma:\n').strip()
-    collection = [int(item) for item in user_input.split(',')]
-    try:
-        __assert_sorted(collection)
-    except ValueError:
-        sys.exit('Sequence must be sorted to apply binary search')
-
-    target_input = raw_input('Enter a single number to be found in the list:\n')
-    target = int(target_input)
-    result = binary_search(collection, target)
-    if result is not None:
-        print('{} found at positions: {}'.format(target, result))
-    else:
-        print('Not found')
+def binary_search(array, target):
+    lower = 0
+    upper = len(array)
+    while lower < upper:   # use < instead of <=
+        x = lower + (upper - lower) // 2
+        val = array[x]
+        if target == val:
+            return x
+        elif target > val:
+            if lower == x:   # these two are the actual lines
+                break        # you're looking for
+            lower = x
+        elif target < val:
+            upper = x
```

Comment:
@shawabhishek Remeber a good code is that which have well-formatted comments, unit test, and must be understood by anyone may be non-programmer. I guess you get the point what I want to say.
Thanks

Reply:
I understand. thanks for guiding me.